crypto/curve25519/curve25519.c

```diff
@@ -31,33 +31,223 @@
+//
+// For x86_64, |x25519_s2n_bignum_alt_capable| always returns 1. If x25519
+// s2n-bignum capable, the x86_64 s2n-bignum-alt version should be supported on
+// pretty much any x86_64 CPU.
+//
+// For all other architectures, return 0.
+OPENSSL_INLINE int x25519_s2n_bignum_alt_capable(void) {
+#if defined(OPENSSL_X86_64)
+  return 1;
+#elif defined(OPENSSL_AARCH64)
+  if (CRYPTO_is_ARMv8_wide_multiplier_capable() == 1) {
+    return 1;
+  } else {
+    return 0;
+  }
+#else
+  return 0;
 #endif
+}
+
+// For aarch64, |x25519_s2n_bignum_alt_capable| always returns 1. If x25519
+// s2n-bignum capable, the Armv8 s2n-bignum-alt version should be supported on
+// pretty much any Armv8 CPU.
+//
+// For x86_64, |x25519_s2n_bignum_alt_capable| returns 1 if we detect support
+// for bmi+adx instruction sets. Return 0 otherwise.
+//
+// For all other architectures, return 0.
+OPENSSL_INLINE int x25519_s2n_bignum_no_alt_capable(void) {
+#if defined(OPENSSL_X86_64)
+  if (CRYPTO_is_BMI2_capable() == 1 && CRYPTO_is_ADX_capable() == 1) {
+    return 1;
+  } else {
+    return 0;
+  }
+#elif defined(OPENSSL_AARCH64)
+  return 1;
+#else
+  return 0;
+#endif
+}
+
+
+// Below is the decision logic for which assembly backend implementation
+// of x25519 s2n-bignum we should use if x25519 s2n-bignum capable. Currently,
+// we support the following implementations.
+//
+// x86_64:
+//   - s2n-bignum-no-alt: hardware implementation using bmi2+adx instruction sets
+//   - s2n-bignum-alt: hardware implementation using standard instructions
+//
+// aarch64:
+//   - s2n-bignum-no-alt: hardware implementation for "small" multiplier throughput
+//   - s2n-bignum-alt: hardware implementation for "higher" multiplier throughput
+//
+// Through experiments we have found that:
+//
+// For x86_64: bmi+adc will almost always give a performance boost. So, here we
+//   prefer s2n-bignum-no-alt over s2n-bignum-alt if the former is supported.
+// For aarch64: if a wide multiplier is supported, we prefer s2n-bignum-alt over
+//   s2n-bignum-no-alt if the former is supported.
+//   x25519_s2n_bignum_alt_capable() specifically looks to match CPUs that have
+//   wide multipliers. this ensures that s2n-bignum-alt will only be used on
+//   such CPUs.
+
+static void x25519_s2n_bignum(uint8_t out_shared_key[32],
+  const uint8_t private_key[32], const uint8_t peer_public_value[32]) {
+
+  uint8_t private_key_internal_demask[32];
+  OPENSSL_memcpy(private_key_internal_demask, private_key, 32);
+  private_key_internal_demask[0] &= 248;
+  private_key_internal_demask[31] &= 127;
+  private_key_internal_demask[31] |= 64;
+
+#if defined(OPENSSL_X86_64)
+
+  if (x25519_s2n_bignum_no_alt_capable() == 1) {
+    curve25519_x25519_byte(out_shared_key, private_key_internal_demask,
+      peer_public_value);
+  } else if (x25519_s2n_bignum_alt_capable() == 1) {
+    curve25519_x25519_byte_alt(out_shared_key, private_key_internal_demask,
+      peer_public_value);
+  } else {
+    abort();
+  }
```

Comment:
Is it possible to get here? Up above each of the functions have a guaranteed supported path between alt and non-alt. I prefer the p384 style of branching just on using alt or not.

Reply:
We could add a total fallback. That's possible.

Well, in theory this code here is more robust then p{384,521}. Not that I think that p{384,521} would ever reasonably fail.

Reply:
I did this. Not sure I like it(?)

Reply:
discussed offline: falling back to C-implementation would basically imply we are doing something very invalid e.g. running a 64-bit compiled program on a 32-bit architecture. This wouldn't work anyway.